File: app/src/main/java/ie/napkin/supertasks/data/format/Markdown.kt
package ie.napkin.supertasks.data.format

/**
 * Where the app agrees with itself about what `**bold**` means.
 *
 * Emphasis lives in the stored text — a block is an arbitrary string, so the markers simply sit in
 * it, stay greppable, survive export and sync, and are what you typed. That makes *reading* them a
 * property of the format rather than of any one screen, which is why the rules are here and not
 * beside a `SpanStyle`.
 *
 * **Prose only.** A note, a heading, a bullet — things written to be read. A task title is not prose
 * but a name, and it is deliberately literal: it appears on surfaces that cannot style anything at
 * all (a widget, a notification, the archive, the focus screen), where emphasis could only be
 * dropped or shown raw. Rather than have a title mean one thing on a page and another in a widget,
 * it means exactly the characters it contains, everywhere.
 *
 * Rendering keeps the markers and merely dims them, so the string shown is exactly as long as the
 * string stored and the caret cannot drift.
 */
object Markdown {

    enum class Kind { CODE, BOLD, ITALIC, BOLD_ITALIC }

    /** One run of emphasis: [outer] includes the markers, [inner] is the text between them. */
    data class Run(val kind: Kind, val outer: IntRange, val inner: IntRange)

    /**
     * `***both***` needs its own pattern and has to be tried first.
     *
     * The bold pattern would otherwise match `***both**` — two markers at the front, two at the back
     * — leaving a stray asterisk undimmed and the closing pair unmatched.
     */
    private val BOLD_ITALIC = Regex("""\*\*\*(?=\S)(.+?)(?<=\S)\*\*\*""")
    private val BOLD = Regex("""\*\*(?=\S)(.+?)(?<=\S)\*\*""")
    private val ITALIC = Regex("""(?<!\*)\*(?=\S)([^*]+?)(?<=\S)\*(?!\*)""")
    private val CODE = Regex("""`(?=\S)([^`]+?)(?<=\S)`""")

    /**
     * Every emphasis run in [text], nested ones included, in the order they start.
     *
     * Nesting is why this recurses: `**a *b* c**` has an italic run inside a bold one, and seeing
     * only the outer run would leave the inner markers undimmed and the inner words unitalicised.
     *
     * Code is the one thing not recursed into: inside backticks an asterisk is content, and
     * `` `2 * 3` `` must survive intact.
     */
    fun runs(text: String): List<Run> = runsIn(text, 0, text.length).sortedBy { it.outer.first }

    private fun runsIn(text: String, from: Int, to: Int): List<Run> {
        if (to - from < 3) return emptyList()
        val slice = text.substring(from, to)
        val found = mutableListOf<Run>()
        // Claimed characters, so `*italic*` cannot match the inner asterisks of `**bold**`, and code
        // spans win over anything that looks like emphasis inside them.
        val claimed = BooleanArray(slice.length)

        fun scan(pattern: Regex, markerLength: Int, kind: Kind) {
            for (match in pattern.findAll(slice)) {
                val outer = match.range
                if (outer.any { claimed[it] }) continue
                val innerStart = outer.first + markerLength
                val innerEnd = outer.last + 1 - markerLength
                if (innerEnd <= innerStart) continue
                outer.forEach { claimed[it] = true }
                found += Run(
                    kind,
                    (from + outer.first)..(from + outer.last),
                    (from + innerStart) until (from + innerEnd),
                )
                // Inside a code span the markers are content; anywhere else they are markers.
                if (kind != Kind.CODE) {
                    found += runsIn(text, from + innerStart, from + innerEnd)
                }
            }
        }

        // Longest marker first: *** before **, ** before *, or each pattern eats part of a longer
        // pair and leaves the remainder stranded.
        scan(CODE, 1, Kind.CODE)
        scan(BOLD_ITALIC, 3, Kind.BOLD_ITALIC)
        scan(BOLD, 2, Kind.BOLD)
        scan(ITALIC, 1, Kind.ITALIC)
        return found
    }
}

File: web/src/format/markdown.ts
/**
 * Inline emphasis — a port of `data/format/Markdown.kt`.
 *
 * Not a markdown parser. It finds the runs of emphasis in one line and says where their markers and
 * their words are, which is what both a renderer and a marker-stripper need. Block structure is the
 * page codec's job; this only ever looks at the inside of a line.
 */

export type Kind = 'code' | 'bold' | 'italic' | 'boldItalic'

/** One run: [outerStart, outerEnd) includes the markers, [innerStart, innerEnd) is the text. */
export interface Run {
  kind: Kind
  outerStart: number
  outerEnd: number
  innerStart: number
  innerEnd: number
}

/**
 * `***both***` needs its own pattern and has to be tried first: the bold pattern would otherwise
 * match `***both**` — two markers at the front, two at the back — leaving a stray asterisk and an
 * unmatched closing pair.
 */
const BOLD_ITALIC = /\*\*\*(?=\S)([\s\S]+?)(?<=\S)\*\*\*/g
const BOLD = /\*\*(?=\S)([\s\S]+?)(?<=\S)\*\*/g
const ITALIC = /(?<!\*)\*(?=\S)([^*]+?)(?<=\S)\*(?!\*)/g
const CODE = /`(?=\S)([^`]+?)(?<=\S)`/g

/** Every emphasis run in [text], nested ones included, in the order they start. */
export function runs(text: string): Run[] {
  return runsIn(text, 0, text.length).sort((a, b) => a.outerStart - b.outerStart)
}

function runsIn(text: string, from: number, to: number): Run[] {
  if (to - from < 3) return []
  const slice = text.slice(from, to)
  const found: Run[] = []
  // Claimed characters, so `*italic*` cannot match the inner asterisks of `**bold**`, and code
  // spans win over anything that looks like emphasis inside them.
  const claimed = new Array<boolean>(slice.length).fill(false)

  const scan = (pattern: RegExp, markerLength: number, kind: Kind) => {
    pattern.lastIndex = 0
    for (const match of slice.matchAll(pattern)) {
      const start = match.index
      const end = start + match[0].length // exclusive
      let taken = false
      for (let i = start; i < end; i++) if (claimed[i]) { taken = true; break }
      if (taken) continue
      const innerStart = start + markerLength
      const innerEnd = end - markerLength
      if (innerEnd <= innerStart) continue
      for (let i = start; i < end; i++) claimed[i] = true
      found.push({
        kind,
        outerStart: from + start,
        outerEnd: from + end,
        innerStart: from + innerStart,
        innerEnd: from + innerEnd,
      })
      // Inside a code span the markers are content; anywhere else they are markers.
      if (kind !== 'code') found.push(...runsIn(text, from + innerStart, from + innerEnd))
    }
  }

  // Longest marker first: *** before **, ** before *, or each pattern eats part of a longer pair
  // and leaves the remainder stranded.
  scan(CODE, 1, 'code')
  scan(BOLD_ITALIC, 3, 'boldItalic')
  scan(BOLD, 2, 'bold')
  scan(ITALIC, 1, 'italic')
  return found
}

/**
 * [text] with its emphasis markers taken out and nothing put in their place.
 *
 * Only the outermost run of a nesting is unwrapped, matching what a renderer does: a surface that
 * strips more than the renderer would is showing a different title from the one beside it.
 */
export function plain(text: string): string {
  const rs = runs(text)
  if (rs.length === 0) return text
  let out = ''
  let at = 0
  let lastEnd = -1
  for (const r of rs) {
    if (r.outerStart < lastEnd) continue // nested inside one already unwrapped
    out += text.slice(at, r.outerStart)
    out += text.slice(r.innerStart, r.innerEnd)
    at = r.outerEnd
    lastEnd = at
  }
  return out + text.slice(at)
}

/** A rendering tree: nested spans with their emphasis, markers removed. */
export type Span = { text: string } | { kind: Kind; children: Span[] }

/**
 * [text] as spans ready to draw.
 *
 * **Only the outermost run of a nesting is unwrapped, and the inner one keeps its markers.** That
 * looks like a shortcut and is the opposite: `InlineText.stripEmphasis` on Android does exactly
 * this, so `**a *b* c**` is drawn bold with the asterisks around `b` still visible, and
 * [plain] leaves those same asterisks behind. A renderer that went one level deeper would show a
 * different string from the one the stripper produces for a widget or a notification — and the two
 * sitting side by side disagreeing about a task's name is the failure both are written to avoid.
 *
 * The last test in the suite pins that: what this draws is character-for-character what [plain]
 * returns.
 */
export function spans(text: string): Span[] {
  const out: Span[] = []
  let at = 0
  let lastEnd = -1
  for (const r of runs(text)) {
    if (r.outerStart < lastEnd) continue // nested inside one already unwrapped
    if (r.outerStart > at) out.push({ text: text.slice(at, r.outerStart) })
    out.push({ kind: r.kind, children: [{ text: text.slice(r.innerStart, r.innerEnd) }] })
    at = r.outerEnd
    lastEnd = at
  }
  if (at < text.length) out.push({ text: text.slice(at) })
  return out
}
